File: JS/Medium/BinaryTree/postorderTraversal/lowestCommonAncestor.js
/**236
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
  let count = 0;
  let lca = null;
  let postOrder = (curr) => {
    if (!curr || lca) return 0;

    let leftCount = postOrder(curr.left);
    let rightCount = postOrder(curr.right);

    if (curr.val == p.val || curr.val == q.val) {
      ++count;
    }

    count = count + leftCount + rightCount;

    if (count == 2 && !lca) {
      lca = curr;
    }

    return count;
  };
  postOrder(root);
  return;
};

var lowestCommonAncestor = function (root, p, q) {
  if (!root || root === p || root === q) return root;

  let left = lowestCommonAncestor(root.left, p, q);
  let right = lowestCommonAncestor(root.right, p, q);

  if (left && right) return root; // p and q found in both sides → root is LCA
  return left || right; // otherwise return whichever is found
};

File: C#/DSA/Trees/LowestCommonAncestor.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class SolutionT
{
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q)
    {
        //Let's do DFS bottom up and if we find p or q? that should return 1
        //else all other should return 0
        //so when ever we get 2 that means we found both p and q , then will determine LCA
        TreeNode ans = null;
        int DFS(TreeNode curr)
        {
            if (ans != null) return -1; //already found, prune recursion
            if (curr == null) return 0;

            int left = DFS(curr.left);
            int right = DFS(curr.right);

            int mid = (curr == p || curr == q) ? 1 : 0;
            if (left + right + mid == 2 && ans == null) ans = curr;

            return left + right + mid;

        }
        DFS(root);
        return ans;
    }
}